### PR TITLE
Add coverage workflow

### LINKED ISSUE #683: Add code coverage and run it on CI
We can borrow from hotshot

- https://github.com/EspressoSystems/HotShot/blob/main/.github/workflows/coverage.yml

### PR DESCRIPTION
**Pull request by tbro**
Adds coverage.yaml and ads coverage shell to flake. also addes code-coverage command to  `justfile`. 

Closes #683

### CODE REVIEW

File: .github/workflows/coverage.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      - name: Checkout Repository
+        uses: actions/checkout@v4
+
+      - name: Install just command runner
+        run: |
+          sudo snap install --edge --classic just
+          just --version
+
+      - name: Generate coverage reports
+        # Use the `release` profile rather than `release-lto` as other workflows do, since `--
```
Comment (sveitser) [inline .github/workflows/coverage.yml:39]: I think we don't use `release-lto`

File: .github/workflows/coverage.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+env:
+  RUSTFLAGS: "--cfg async_executor_impl=\"async-std\" --cfg async_channel_impl=\"async-std\""
+
+jobs:
+  code-coverage:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Install Nix
+        uses: cachix/install-nix-action@v25
```
Comment (sveitser) [inline .github/workflows/coverage.yml:21]: This job looks like it doesn't use `nix`.

**Comment by sveitser** [comment]
I think we can definitely not run this on every PR. I'm trying to run it locally but it's taking ages.

For the CI to work we need to either run it with nix (then you get all the dependencies) or install some missing dependencies (at least foundry).

**Comment by sveitser** [comment]
I have these two tests failing locally

```
failures:
    api::fs::generic_tests::state_signature_test_with_query_module
    api::sql::generic_tests::state_signature_test_with_query_module

test result: FAILED. 47 passed; 2 failed; 2 ignored; 0 measured; 0 filtered out; finished in 991.56s
```

**Comment by sveitser** [comment]
If I run them individually with this coverage configuration they both pass relatively quickly (about 10s) but running together they take 100 seconds. :thinking:

**Comment by sveitser** [comment]
Hotshot actually runs the coverage workflow with tokio, not async_std.

**Comment by tbro** [comment]
> Hotshot actually runs the coverage workflow with tokio, not async_std.

Yes, I've tried both w/out success. I'm just trying things :crossed_fingers:

File: justfile [hunk 3]
```diff
@@ -83,3 +86,11 @@ dev-deploy url="http://localhost:8545" mnemonics="test test test test test test
...
     --sig "run(uint32 numBlocksPerEpoch, uint32 numInitValidators)" {{num_blocks_per_epoch}} {{num_init_validators}} \
     --fork-url {{url}} --broadcast
+
+@tokio target *ARGS:
+  echo setting executor to tokio
+  export RUSTDOCFLAGS='-D warnings --cfg async_executor_impl="tokio" --cfg async_channel_impl="tokio" {{original_rustdocflags}}' RUSTFLAGS='--cfg async_executor_impl="tokio" --cfg async_channel_impl="tokio" {{original_rustflags}}' && just {{target}} {{ARGS}}
+
+code-coverage:
+  echo "Running code coverage"
+  cargo-llvm-cov llvm-cov --lib --bins --tests --release --lcov --output-path lcov.info -p sequencer -- --skip sql
```
Comment (sveitser) [inline justfile:96]:
```suggestion
    nix develop .#nightly -c cargo tarpaulin --verbose --all-features --workspace --timeout 120 --out xml
```

File: .github/workflows/coverage.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
+name: Code Coverage Workflow
+
+on:
+  schedule:
+    - cron: '0 0 * * 1'
+  pull_request:
```
Comment (sveitser) [inline .github/workflows/coverage.yml:6]:
```suggestion
  push:
    branches:
      - main
      - release-*
```

Suggest not running this on PRs. Manual runs can still be triggers in the actions page thanks to `workflow_dispatch`.

File: justfile (hunk 3 above)
Comment (sveitser) [reply justfile:95]: This suggestion would only apply if we were using tarpaulin.

File: .github/workflows/coverage.yml (hunk 4 above)
Comment (tbro) [reply .github/workflows/coverage.yml:6]: Ok. I'll make this change before we merge... For now we need it to run on this PR to test it.

File: justfile [hunk 5]
```diff
@@ -1,6 +1,9 @@
 default:
     just --list
 
+original_rustflags := env_var_or_default('RUSTFLAGS', '--cfg hotshot_example')
+original_rustdocflags := env_var_or_default('RUSTDOCFLAGS', '--cfg hotshot_example')
+
```
Comment (sveitser) [inline justfile:6]:
```suggestion
```

Comment (sveitser) [reply justfile:6]: Looks like no longer used.

File: flake.nix [hunk 6]
```diff
@@ -193,6 +193,7 @@
             cargo-audit
             cargo-edit
             cargo-sort
+            cargo-llvm-cov
```
Comment (sveitser) [inline flake.nix:196]: This we can also remove since we have a separate dev shell for coverage now.

File: .github/workflows/coverage.yml [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+jobs:
+  code-coverage:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v4
+
+      - name: Enable Rust Caching
+        uses: Swatinem/rust-cache@v2
```
Comment (sveitser) [inline .github/workflows/coverage.yml:24]: According to https://github.com/marketplace/actions/rust-cache (first comment in example) the cache should come after the installation of the rust toolchain.

**Comment by sveitser** [comment]
@tbro roughly how long does it take for you when you run this locally? The test `snark::tests::test_proof_generation` has been running for multiple minutes.

File: flake.nix [hunk 8]
```diff
@@ -253,6 +254,28 @@
...
+          };
+        in
+        mkShell {
+          buildInputs = [
+            # Rust dependencies
+            pkg-config
+            openssl
+            curl
+            protobuf # to compile libp2p-autonat
+            toolchain
```
Comment (sveitser) [inline flake.nix:270]:
```suggestion
            toolchain
            grcov
```

File: justfile [hunk 9]
```diff
@@ -82,3 +82,11 @@ dev-deploy url="http://localhost:8545" mnemonics="test test test test test test
...
     --sig "run(uint32 numBlocksPerEpoch, uint32 numInitValidators)" {{num_blocks_per_epoch}} {{num_init_validators}} \
     --fork-url {{url}} --broadcast
+
+# This is meant for local development and produces HTML output. In CI
+# the lcov output is pushed to coveralls.
+code-coverage:
+  echo "Running code coverage"
+  nix develop .#coverage -c cargo test --all-features --no-fail-fast --release --workspace -- --skip service::test::test_
+  grcov . -s . --binary-path $CARGO_TARGET_DIR/debug/ -t html --branch --ignore-not-existing -o $CARGO_TARGET_DIR/coverage/
+  open $CARGO_TARGET_DIR/coverage/index.html
```
[the remark this answers is not in the document]
Comment (tbro) [reply justfile:92]: Shoot I thought I had found a command that would work across linux/mac. In that case I think its probably better to just output a message containing the location of the HTML, and the caller can open it as they like.

**Comment by sveitser** [comment]
```
[![Coverage Status](https://coveralls.io/repos/github/EspressoSystems/espresso-sequencer/badge.svg?branch=main)](https://coveralls.io/github/EspressoSystems/espresso-sequencer?branch=main)
```
Should do the trick

**Comment by tbro** [comment]
> @tbro roughly how long does it take for you when you run this locally? The test `snark::tests::test_proof_generation` has been running for multiple minutes.

I haven't timed it yet, but it does complete which is more than it used to do. The `proof_generation` test takes a long time, but it does eventually succeed.

File: justfile (hunk 9 above)
[the remark this answers is not in the document]
Comment (tbro) [reply justfile]: fixed 407d66bc1db0417baecff6abf9203e4fae088290

File: flake.nix (hunk 8 above)
Comment (tbro) [reply flake.nix:267]: fixed 50119481eb481933ac3bf71b275fae8868e9d524

File: flake.nix [hunk 10]
```diff
@@ -253,6 +254,28 @@
           ];
           inherit RUST_LOG RUST_BACKTRACE RUSTFLAGS CARGO_TARGET_DIR;
         };
+      devShells.coverage =
+        let
+          toolchain = pkgs.rust-bin.nightly.latest.minimal.override {
```
[the remark this answers is not in the document]
Comment (tbro) [reply flake.nix]: fixed 50119481eb481933ac3bf71b275fae8868e9d524

File: .github/workflows/coverage.yml (hunk 7 above)
Comment (tbro) [reply .github/workflows/coverage.yml]: 49cafc31c94ca6ff54f5da806c8b9df670bd75b2

**Comment by tbro** [comment]
> ```
> [Coverage Status](https://coveralls.io/github/EspressoSystems/espresso-sequencer?branch=main)
> ```
> 
> Should do the trick

added the badge in 49cafc31c94ca6ff54f5da806c8b9df670bd75b2

**Comment by sveitser** [comment]
@tbro I added ignoring of files in `contract-bindings` and `contracts`.

**Review by sveitser — APPROVED** [review]
LGTM 🥳

**Pull request merged**